ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/assertions/CommandDeduplicationAssertions.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.infrastructure.assertions
+
+import java.time.Duration
+
+import com.daml.api.util.DurationConversion
+import com.daml.ledger.api.testtool.infrastructure.Assertions.{assertDefined, fail}
+import com.daml.ledger.api.testtool.infrastructure.participant.{
+  CompletionResponse,
+  ParticipantTestContext,
+}
+import com.daml.ledger.api.v1.experimental_features.CommandDeduplicationPeriodSupport.{
+  DurationSupport,
+  OffsetSupport,
+}
+import com.daml.ledger.client.binding.Primitive.Party
+import com.daml.platform.testing.WithTimeout
+import com.google.protobuf.duration.{Duration => DurationProto}
+import io.grpc.Status.Code
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits._
+
+object CommandDeduplicationAssertions {
+
+  def assertDeduplicationDuration(
+      requestedDeduplicationDuration: DurationProto,
+      completionResponse: CompletionResponse,
+      submittingParty: Party,
+      ledger: ParticipantTestContext,
+  )(implicit executionContext: ExecutionContext): Future[Unit] = {
+    val requestedDuration = DurationConversion.fromProto(requestedDeduplicationDuration)
+    ledger.features.commandDeduplicationFeatures.getDeduplicationPeriodSupport.durationSupport match {
+      case DurationSupport.DURATION_NATIVE_SUPPORT =>
+        val reportedDurationProto = assertDefined(
+          completionResponse.completion.deduplicationPeriod.deduplicationDuration,
+          "No deduplication duration has been reported",
+        )
+        val reportedDuration = DurationConversion.fromProto(reportedDurationProto)
+        assert(
+          reportedDuration >= requestedDuration,
+          s"The reported deduplication duration $reportedDuration was smaller than the requested deduplication duration $requestedDuration.",
+        )
+        Future.unit
+      case DurationSupport.DURATION_CONVERT_TO_OFFSET =>
+        val reportedOffset = assertDefined(
+          completionResponse.completion.deduplicationPeriod.deduplicationOffset,
+          "No deduplication offset has been reported",
+        )
+        if (completionResponse.completion.getStatus.code == Code.ALREADY_EXISTS.value) {
+          val completionStreamRequest = ledger.completionStreamRequest()(submittingParty)
```

Comment:
Do I read this correctly saying that this requests the completion stream from ledger begin? If so, the `findCompletion` below will look for all (or maybe just the first?) matching completions. Accordingly, the assertion `acceptedCompletionResponse.offset.getAbsolute >= reportedOffset` below is required for all / the first such completion. If there are several accepting completions (with disjoint deduplication periods) on the ledger, then this assertion is likely to fail spuriously on the first such completion.

I recommend to adjust the `findCompletion` search criterion so that it skips all completions with offsets less than `reportedOffset`.

Reply:
It requests from ledger end queried when creating a test context (for each **test case**). We're using it in other (dedup) test cases. So far, we don't have test cases where there are multiple consecutive accepting completions. What I can do to fix it for the future is also query by a submission ID. WDYT, @andreaslochbihler-da?

Reply:
Then I think it's fine as is because test cases are properly isolated against each other. It's probably not worth the effort to future-proof this right now if we don't immediately see the need for the generalization. You may want to mention this limitation in a comment though so that we save the thinking if we ever run into this case in the future.